src/html-purify.js

```diff
@@ -53,14 +53,56 @@ See the accompanying LICENSE file for terms.
         return false;
     }
 
+    function addQuoteToValue(action, value) {
+        if (action & derivedState.TransitionName.DQ_ATTR) {         // collect double quoted value
+            return '"' + value + '"';
+        } 
+        else if (action & derivedState.TransitionName.SQ_ATTR) {    // collect single quoted value
+            return "'" + value + "'";
+        } 
+        else /*if (action & derivedState.TransitionName.UQ_ATTR) {  // collect unquoted value */{
+            return value;
+        }
+    }
+
     function processTransition(prevState, nextState, i) {
         /* jshint validthis: true */
         /* jshint expr: true */
         var parser = this.parser,
+            action = derivedState.Transitions[prevState][nextState],
             idx, tagName, attrValString, openedTag, key, value;
 
-        
-        switch (derivedState.Transitions[prevState][nextState]) {
+        // check if tag and/or attr is available for collection
+        // collect the values (if any) only when the attr name is allowed
+        if (action & 0xF && contains(this.attributesWhitelist, (key = parser.getAttributeName()))) {
+
+            // collect the attr name only
+            // boolean attributes may not have a value
+            if (action & derivedState.TransitionName.NV_ATTR) {
+                this.attrVals[key] = null;
+            }
+            // collect both the attr name and value
+            else {
+                value = parser.getAttributeValue() || '';
+
+                // store only valid style attribute value
+                if (key === 'style') {
+                    if (this.cssParser.parseCssString(value)) {
+                        this.attrVals[key] = addQuoteToValue(action, value);
+                    }
+                } 
+                // apply the blacklist filter for URI attr value
+                else if (hrefAttribtues[key]) {
+                    this.attrVals[key] = addQuoteToValue(action, uriBlacklistFilter(value));
+                }
+                else {
+                    this.attrVals[key] = addQuoteToValue(action, value);
+                }
+            }
+        }
+
+        // drop the last 4 bits that indicate quoted status
+        switch (action & 0xF0) {
```

Comment:
can we have meaningful macros name instead of 0xF0 here? Something like QUOTE_TYPE_MASK ?